src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp

```diff
@@ -16,577 +16,1263 @@
-                continue;
-
-            if (player->IsAlive())
-            {
-                temp->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
-                temp->SetImmuneToPC(true);
-                temp->SetReactState(REACT_AGGRESSIVE);
-                temp->EngageWithTarget(player);
-            }
+            if (instance->GetGuidData(DATA_GRAND_CHAMPION_1) == me->GetGUID())
+                instance->SetGuidData(DATA_GRAND_CHAMPION_VEHICLE_1, mount->GetGUID());
+            else if (instance->GetGuidData(DATA_GRAND_CHAMPION_2) == me->GetGUID())
+                instance->SetGuidData(DATA_GRAND_CHAMPION_VEHICLE_2, mount->GetGUID());
+            else if (instance->GetGuidData(DATA_GRAND_CHAMPION_3) == me->GetGUID())
+                instance->SetGuidData(DATA_GRAND_CHAMPION_VEHICLE_3, mount->GetGUID());
+            mount->SetOrientation(centerOrientation);
+            mount->SetHomePosition(me->GetHomePosition());
+            return mount;
         }
     }
+    if (me->GetGUID() == instance->GetGuidData(DATA_GRAND_CHAMPION_1))
+        return ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_VEHICLE_1));
+    else if (me->GetGUID() == instance->GetGuidData(DATA_GRAND_CHAMPION_2))
+        return ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_VEHICLE_2));
+    else if (me->GetGUID() == instance->GetGuidData(DATA_GRAND_CHAMPION_3))
+        return ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_VEHICLE_3));
+
+    // should not happen
+    return nullptr;
 }
 
-bool GrandChampionsOutVehicle(Creature* me)
+// called on rider
+void SetGrandChampionToEvadeMode(Creature* me)
 {
-    InstanceScript* instance = me->GetInstanceScript();
-
-    if (!instance)
-        return false;
-
-    Creature* pGrandChampion1 = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_1));
-    Creature* pGrandChampion2 = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_2));
-    Creature* pGrandChampion3 = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_3));
-
-    if (pGrandChampion1 && pGrandChampion2 && pGrandChampion3)
+    me->RemoveAllAuras();
+    me->ApplySpellImmune(SPELL_TRAMPLE_AURA, IMMUNITY_ID, SPELL_TRAMPLE_AURA, false);
+    me->GetThreatManager().ClearAllThreat();
+    me->CombatStop(true);
+    me->SetLootRecipient(nullptr);
+    me->AI()->Reset();
+    if (Creature* mount = FindMyMount(me))
     {
-        if (!pGrandChampion1->HasUnitMovementFlag(MOVEMENTFLAG_ONTRANSPORT) &&
-            !pGrandChampion2->HasUnitMovementFlag(MOVEMENTFLAG_ONTRANSPORT) &&
-            !pGrandChampion3->HasUnitMovementFlag(MOVEMENTFLAG_ONTRANSPORT))
-            return true;
+        mount->AI()->Reset();
+        EnterAggressiveMode(mount);
+        mount->GetMotionMaster()->Clear();
+        mount->GetMotionMaster()->MoveTargetedHome();
     }
-
-    return false;
+    DoCastPennant(me);
```

Comment:
all this pennant logic makes me wonder if all of this couldn't simply be handled by an appropriate entry in `creature_template_addon`'s aura column...